src/lib.rs

```diff
@@ -263,54 +284,54 @@ impl ThreadedClient for Client {
-            listener: listener,
-            read_preference: rp,
-            write_concern: wc,
-            log_file: file,
-        });
-
-        // Fill servers array and set options
-        {
-            let top_description = &client.topology.description;
-            let mut top = try!(top_description.write());
-            top.heartbeat_frequency_ms = client_options.heartbeat_frequency_ms;
-            top.server_selection_timeout_ms = client_options.server_selection_timeout_ms;
-            top.local_threshold_ms = client_options.local_threshold_ms;
-
-            for host in &config.hosts {
-                let server =
-                    Server::new(client.clone(), host.clone(), top_description.clone(), true);
-                top.servers.insert(host.clone(), server);
+        -> Result<Client> {
+
+            let client_options = options.unwrap_or_else(ClientOptions::new);
+
+            let rp = client_options.read_preference
+                .unwrap_or_else(|| ReadPreference::new(ReadMode::Primary, None));
+            let wc = client_options.write_concern.unwrap_or_else(WriteConcern::new);
+
+            let listener = Listener::new();
+            let file = match client_options.log_file {
+                Some(string) => {
+                    let _ = listener.add_start_hook(log_command_started);
+                    let _ = listener.add_completion_hook(log_command_completed);
+                    Some(Mutex::new(try!(OpenOptions::new()
+                                         .write(true)
+                                         .append(true)
+                                         .create(true)
+                                         .open(&string))))
+                }
+                None => None,
+            };
+
+            let client = Arc::new(ClientInner {
+                req_id: Arc::new(ATOMIC_ISIZE_INIT),
+                topology: try!(Topology::new(config.clone(), description, client_options.stream_connector.clone())),
+                listener: listener,
+                read_preference: rp,
+                write_concern: wc,
+                log_file: file,
+            });
+
+            // Fill servers array and set options
+            {
+                let top_description = &client.topology.description;
+                let mut top = try!(top_description.write());
+                top.heartbeat_frequency_ms = client_options.heartbeat_frequency_ms;
+                top.server_selection_timeout_ms = client_options.server_selection_timeout_ms;
+                top.local_threshold_ms = client_options.local_threshold_ms;
+
+                for host in &config.hosts {
+                    let server = Server::new(client.clone(), host.clone(), top_description.clone(), true, client_options.stream_connector.clone());
+
+                    top.servers.insert(host.clone(), server);
+                }
             }
-        }
 
-        Ok(client)
-    }
+            Ok(client)
+        }
```

Comment:
Yeah, I've grown less and less fond of rustfmt after using clang-format quite a bit at work. I'm kind of hoping it matures into something better at some point, but it's not really that close yet.